File: Lab3/src/KI_306/Voitovych/Lab3/Cat.java
/**
 * 
 */
package KI_306.Voitovych.Lab3;

import java.io.FileWriter;
import java.io.IOException;
/**
 * Class <code>Cat</code> implements cat
 */
public abstract class Cat {

	protected String name;
	protected Breed breed;
	protected Master number, nameMaster;
	protected Collar collar;
	protected int energy;
	protected int food;
	protected String location;


    /**
     * Constructor
     */
    public Cat() {
        this.collar = new Collar();
        this.breed = new Breed();
        this.number = new Master();
        this.nameMaster = new Master();
        logActivity("Cat's name: None "
        		+ "Cat's bread: None"+
        		"Cat's master: None, None"+
        		"Cat's colar: None");
    }

    /**
     * Constructor
     * @param <code>name</code> name cats
     * @param <code>breed</code> breed cats
     * @param <code>number</code> number master
     * @param <code>nameMaster</code> name master
     * @param <code>info_collar</code> info_collar
     */
    public Cat(String name, String breed, String number, String nameMaster, boolean info_collar, String location) {
        this.collar = new Collar();
        this.breed = new Breed();
        this.number = new Master();
        this.nameMaster = new Master();
        
    	this.name = name; 
    	this.energy = 5;
    	this.food = 5;
    	this.location = location;
    	this.collar.setPresence(info_collar);
        this.breed.setBreed(breed);
        this.number.setNumber(number);
        this.nameMaster.setNameMaster(nameMaster);
        logActivity("Cat's name: "+name
        		+ ", Cat's bread: "+ breed +
        		", Number master: "+number + ", Name master: "+nameMaster+
        		", Cat's colar: " + info_collar);
    }

	/**
    * Method gets the name cat
    */

    public String getName() {
        return name;
    }
    
    /**
    * Method sets the name cat
    * @param <code>name</code> name cat
    */

    public void setName(String name) {
        this.name = name;
        logActivity("New name for cat's: " + name);
    }
    
    /**
    * Method action cat play
    * @param <code>game</code> game for cat
    */

    public void play(String game) {
    	if (game == "mouse") {
    		this.energy = this.energy - 1;
    		this.food = this.food + 1;
            logActivity(name+ " plays with mouse: energy-1, food+1");
    	}
    	else if(game == "bug") {
    		this.energy = this.energy + 1;
    		this.food = this.food - 1;
    		this.location = "bugs";
    		logActivity(name+ " plays in bug: energy+1, food-1");
    	}
    	else {
    		this.energy = this.energy - 2;
    		this.food = this.food - 1;
    		this.location = "outside";
    		logActivity(name+ " plays in outside: energy-2, food-1");
    	}
    }
    
    /**
    * Method action cat sleep
    */

    public void sleep() {
    	this.energy = this.energy +1;
        logActivity(name + " sleeps: energy+1");
    }
    
    /**
    * Method action cat clean
    */

    public void clean() {
    	this.energy = this.energy - 1;
        logActivity(name + " cleans: energy-1");
    }
    
    /**
    * Method action cat night_vision
    * @param <code>visor</code> on or off night vision
    */

    public void night_vision(boolean visor) {
    	this.energy = this.energy - 1;
        logActivity(name + " night vision "+visor+": energy-1");
    }
    
    /**
    * Method action cat mew
    */

    public abstract void mew();
    
    /**
    * Method location cat
    */

    public void setPlace() {
        logActivity(name + " in " + location);
    }
        
   
    /**
    * Method action cat eat
    */

    public void eat(String food) {
    	if (food == "fish") {
    		this.food = this.food + 1;
            logActivity(name+ " eats fish: food+1");
    	}
    	else if(food == "meat") {
    		this.food = this.food + 2;
    		logActivity(name+ " eats meat: food+2");
    	}
    	else {
    		this.food = this.food - 1;
    		logActivity(name+ " don't want to eat candy: food-1");
    	}
    }
    
    /**
    * Method cat status
    */

    public void status() {
    	if (this.energy == 5 && this.food == 5) {
    		logActivity("Status: Nice");
    	}
    	else if (this.energy < 5 || this.food < 5) {
    		logActivity("Status: "+name+" need to sleep or eat");
    	}
    	else {
    		logActivity("Status: "+name+" want to play");
    	}
    	logActivity("Level energy: " + this.energy+ ", Level food: " + this.food);
    }
    
    
    /**
    * Method information about cat
    */

    public void info_cat() {
        logActivity("Cat's name: "+name
        		+ ", Cat's bread: "+ breed.getBreed() +
        		", Number master: "+ number.getNumber() + ", Name master: "+nameMaster.getNameMaster()+
        		", Cat's colar: " + collar.getPresence() + ", Cat's energy: " + energy);
    }
        
    /**
    * Method logging info
    */

    public void logActivity(String message) {
        try (FileWriter fw = new FileWriter("cat_activity.log", true)) {
            fw.write(message + "\n");
            System.out.println(message); 
        } catch (IOException e) {
            System.err.println("Error: " + e.getMessage());
        }
    }
}

/**
 *  Class describing a breed of cat's
 */
class Breed {
    private String breed;
    
    /**
    * Method gets the name of breed
    */

    public String getBreed() {
        return breed;
    }
    /**
    * Method sets the name of room
    * @param <code>breed</code> name of room
    */
    public void setBreed(String breed) {
        this.breed = breed;
    }
}

/**
 *  Class describing a collar of cat's
 */
class Collar {
    private boolean collar;
    
    /**
    * Method gets the presence of a collar
    */

    public boolean getPresence() {
        return collar;
    }
    /**
    * Method sets the presence of a collar
    * @param <code>collar</code> the presence of a collar
    */

    public void setPresence(boolean collar) {
        this.collar = collar;
    }
}

/**
 *  Class describing a master of cat's
 */
class Master {
    private String nameMaster;
    private String number;
    
    /**
    * Method sets the presence of a collar
    * @param <code>collar</code> the presence of a collar
    */

    public String getNameMaster() {
        return nameMaster;
    }
    
    /**
    * Method sets the presence of a collar
    * @param <code>collar</code> the presence of a collar
    */

    public void setNameMaster(String nameMaster) {
        this.nameMaster = nameMaster;
    }
    
    /**
    * Method gets the number master
    */
    
    public String getNumber() {
        return number;
    }
    
    /**
    * Method sets the number master
    * @param <code>number</code> number master
    */

    public void setNumber(String number) {
        this.number = number;
    }
}

File: Lab9/cat.py
from breed import Breed
from master import Master
from collar import Collar

#Class Cat implements cat
class Cat:
    #Constructor
    def __init__(self, name="None", breed="None", number="None", name_master="None", info_collar=False, location="None"):
        self.name = name
        self.breed = Breed(breed)
        self.master = Master(number, name_master)
        self.collar = Collar(info_collar)
        self.energy = 5
        self.food = 5
        self.location = location
        self.log_activity(f"Cat's name: {self.name}, Cat's breed: {self.breed.get_breed()}, Cat's master: {self.master.get_number()}, {self.master.get_name_master()}, Cat's collar: {self.collar.get_presence()}")

    # Method gets name cat
    def get_name(self):
        return self.name

    # Method sets name cat
    def set_name(self, name):
        self.name = name
        self.log_activity(f"New name for cat's: {self.name}")

    # Method action play cat
    def play(self, game):
        if game == "mouse":
            self.energy -= 1
            self.food += 1
            self.log_activity(f"{self.name} plays with mouse: energy-1, food+1")
        elif game == "bug":
            self.energy += 1
            self.food -= 1
            self.location = "bugs"
            self.log_activity(f"{self.name} plays in bug: energy+1, food-1")
        else:
            self.energy -= 2
            self.food -= 1
            self.location = "outside"
            self.log_activity(f"{self.name} plays outside: energy-2, food-1")

    # Method action sleep cat
    def sleep(self):
        self.energy += 1
        self.log_activity(f"{self.name} sleeps: energy+1")

    # Method action clean cat
    def clean(self):
        self.energy -= 1
        self.log_activity(f"{self.name} cleans: energy-1")

    # Method to activate a night_vision cat's
    def night_vision(self, visor):
        self.energy -= 1
        self.log_activity(f"{self.name} night vision {visor}: energy-1")

    # Method to change location cat's
    def set_place(self):
        self.log_activity(f"{self.name} in {self.location}")

    # Method action eat cat
    def eat(self, food):
        if food == "fish":
            self.food += 1
            self.log_activity(f"{self.name} eats fish: food+1")
        elif food == "meat":
            self.food += 2
            self.log_activity(f"{self.name} eats meat: food+2")
        else:
            self.food -= 1
            self.log_activity(f"{self.name} doesn't want to eat candy: food-1")

    # Method to show status cat's
    def status(self):
        if self.energy == 5 and self.food == 5:
            self.log_activity("Status: Nice")
        elif self.energy < 5 or self.food < 5:
            self.log_activity(f"Status: {self.name} needs to sleep or eat")
        else:
            self.log_activity(f"Status: {self.name} wants to play")
        self.log_activity(f"Energy level: {self.energy}, Food level: {self.food}")

    # Method to show information about cat's
    def info_cat(self):
        self.log_activity(f"Cat's name: {self.name}, Cat's breed: {self.breed.get_breed()}, Master's number: {self.master.get_number()}, Master's name: {self.master.get_name_master()}, Cat's collar: {self.collar.get_presence()}, Cat's energy: {self.energy}")

    # Method to write action in file
    def log_activity(self, message):
        with open("cat_activity.log", "a") as f:
            f.write(message + "\n")
            print(message)